ingesters/SimpleRelay/regexHandlers.go

```diff
@@ -0,0 +1,327 @@
+			if strings.Contains(err.Error(), "closed") {
+				break
+			}
+			failCount++
+			lg.Info("failed to accept connection", log.KV("readertype", `regex`), log.KV("mode", tp.String()), log.KVErr(err))
+			if failCount > 3 {
+				break
+			}
+			continue
+		}
+		debugout("Accepted %v connection from %s in regex mode\n", tp.String(), conn.RemoteAddr())
+		lg.Info("accepted connection", log.KV("address", conn.RemoteAddr()), log.KV("readertype", `regex`), log.KV("mode", tp), log.KV("listener", cfg.name))
+		failCount = 0
+		go regexConnHandler(conn, cfg, igst)
+	}
+	return
+}
+
+func regexConnHandler(c net.Conn, cfg regexHandlerConfig, igst *ingest.IngestMuxer) {
+	cfg.wg.Add(1)
+	id := addConn(c)
+	defer cfg.wg.Done()
+	defer delConn(id)
+	defer c.Close()
+	var rip net.IP
+	var ts entry.Timestamp
+	var tg *timegrinder.TimeGrinder
+	var ok bool
+	var regex *regexp.Regexp
+	var err error
+
+	if regex, err = regexp.Compile(cfg.regex); err != nil {
+		// will never happen (we always check the regex first)
+		return
+	}
+	prefixIndex := regex.SubexpIndex("prefix")
+	suffixIndex := regex.SubexpIndex("suffix")
+
+	if cfg.src == nil {
+		ipstr, _, err := net.SplitHostPort(c.RemoteAddr().String())
+		if err != nil {
+			lg.Error("failed to get host from remote addr", log.KV("remoteaddress", c.RemoteAddr().String()), log.KVErr(err))
+			return
+		}
+		if rip = net.ParseIP(ipstr); rip == nil {
+			lg.Error("failed to get remote address", log.KV("remoteaddress", ipstr))
+			return
+		}
+	} else {
+		rip = cfg.src
+	}
+
+	if !cfg.ignoreTimestamps {
+		var err error
+		tcfg := timegrinder.Config{
+			EnableLeftMostSeed: true,
+		}
+		tg, err = timegrinder.NewTimeGrinder(tcfg)
+		if err != nil {
+			lg.Error("failed to get a handle on the timegrinder", log.KVErr(err))
+			return
+		} else if err = cfg.timeFormats.LoadFormats(tg); err != nil {
+			lg.Error("failed to load custom time formats", log.KVErr(err))
+			return
+		}
+		if cfg.setLocalTime {
+			tg.SetLocalTime()
+		}
+		if cfg.timezoneOverride != `` {
+			err = tg.SetTimezone(cfg.timezoneOverride)
+			if err != nil {
+				lg.Error("failed to set timezone", log.KV("timezone", cfg.timezoneOverride), log.KVErr(err))
+				return
+			}
+		}
+		if cfg.formatOverride != `` {
+			if err = tg.SetFormatOverride(cfg.formatOverride); err != nil {
+				lg.Error("Failed to load format override", log.KV("override", cfg.formatOverride), log.KVErr(err))
+				return
+			}
+		}
+	}
+
+	rd := make([]byte, 1024)
+	bio := bufio.NewReader(c)
+	var buf bytes.Buffer // we read from the connection into this buffer
+	var data []byte
+	var prefix, suffix []byte
+	var done bool
+	var n int
+	for !done {
+		// we build the entry data into this buffer.
+		// make a new one on each loop so we're safe to call Bytes() at the end
+		// read a bunch of bytes off the connection
+		if n, err = bio.Read(rd); err != nil && err != io.EOF {
+			lg.Error("error reading from regex connection", log.KVErr(err))
+			return
```

Comment:
if you have half an entry in the buffer and then get a timeout we lose that half buffered entry, seems like we should throw it.  Same for if you have a complete entry and are waiting on the next delimited entry and get a timeout or network break.  It would lose that buffered entry.

Seems like any "exit" should attempt to process the existing buffer